# AoC_Day8
It seems like you're meant to use the left/right instructions to navigate the network. Perhaps if you have the camel follow the same instructions, you can escape the haunted wasteland!
After examining the maps for a bit, two nodes stick out: AAA and ZZZ. You feel like AAA is where you are now, and you have to follow the left/right instructions until you reach ZZZ.
This format defines each node of the network individually. For example:

```
RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)
```

Starting with AAA, you need to look up the next element based on the next left/right instruction in your input. In this example, start with AAA and go right (R) by choosing the right element of AAA, CCC. Then, L means to choose the left element of CCC, ZZZ. By following the left/right instructions, you reach ZZZ in 2 steps.

Of course, you might not find ZZZ right away. If you run out of left/right instructions, repeat the whole sequence of instructions as necessary: RL really means RLRLRLRLRLRLRLRL... and so on. For example, here is a situation that takes 6 steps to reach ZZZ:

```
LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)
```
Starting at AAA, follow the left/right instructions. How many steps are required to reach ZZZ?

In [ ]:
sample = '''LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)
'''
sample

'LLR\n\nAAA = (BBB, BBB)\nBBB = (AAA, ZZZ)\nZZZ = (ZZZ, ZZZ)\n'

In [ ]:
from aocd import get_data
inp = get_data(day=8, year=2023)
inp[:1000]

'LRLRLLRRLLRRLRRLRRRLLRLRLRLRLRRLRRRLRLRRLRLLRRLLRLRRLRLRRLLRRRLRLRLRRRLRLLRRRLLLLLLRRRLRRLLLRRLRLRRLRRLRLRRLRRLLRRLRRRLRRRLLRLRLLLRRLLLRRLLRRLRLLRRRLRRRLRRRLRLRRLRRLLLRRRLRRLLRRLRRRLRLRLRRLRRLRRRLRRRLRLLLLRRRLRLRRRLRRRLLRLRRLRRLLRLLLRRLRLRRLRRRLRRRLRRRLLRRRLRLLRRRLRRRLRRRLRRRLRRLRRRLLRRLLRLRLRRRLRRRLRLRRRR\n\nLFM = (PCJ, GQH)\nTHP = (SPP, MKB)\nPBL = (VVX, JMM)\nDSR = (BST, MPD)\nQKF = (DSR, KGX)\nAAA = (SLH, CVN)\nDSH = (DVN, VXT)\nVTK = (NSF, QJM)\nQQR = (FNB, CBM)\nRQV = (XQQ, NQQ)\nCGV = (CCB, FTX)\nRXL = (HXQ, HLQ)\nMQP = (MSH, TJB)\nMPL = (JSJ, TJR)\nXQK = (RQT, JKB)\nGXT = (JFP, MTT)\nLHK = (BVR, MFH)\nBXH = (PGR, NDC)\nFFV = (MJC, TLD)\nTFM = (DKJ, RTC)\nMVG = (TNC, PXR)\nQXN = (CGV, KKM)\nLTN = (MQC, XGD)\nNLP = (SQQ, DKG)\nQTK = (MBH, GJX)\nSSR = (DHV, NHN)\nDGJ = (QSX, CKB)\nSFD = (GRS, DLD)\nHBT = (NVN, HJF)\nVKH = (CBT, CBT)\nVRH = (FVN, JNB)\nHTJ = (JNB, FVN)\nBDL = (SQS, QVX)\nMMC = (MJC, TLD)\nRBL = (VNF, BXD)\nRNM = (HJQ, PJF)\nCTZ = (NCP, PLT)\nJFK = (TPJ, JGX)\nHPJ 

**AI Prompt**

Lets solve AoC Day 8 2023. The netwprk data I can put in a dictionary and then start at AAA and get the next key from its values and so on

**AI Response**

Good thinking! Let's break this down into steps. First, would you like to try creating a dictionary from the sample data? We'll need to:

1. Split the input into the instructions (first line) and the network data
2. Parse the network data into a dictionary where each key points to a tuple of left/right choices

Would you like to try writing the code for this first part? I can give you hints if needed.

Remember, we want something like this structure from the sample:
```python
# Example of what we're aiming for:
network = {
    'AAA': ('BBB', 'BBB'),
    'BBB': ('AAA', 'ZZZ'),
    'ZZZ': ('ZZZ', 'ZZZ')
}
```

Would you like to try writing the code to create this dictionary?

In [ ]:
lines = inp.splitlines()
instructions = lines[0]
network = {}
for line in lines[2:]:
    k,v = line.split('=')
    k = k.strip()
    v = v.strip()
    v1 = v[1:4]
    v2 = v[6:9]
    network[k]=(v1,v2)

k = 'AAA'
i = 0
steps = 0
max_i=len(instructions)
while k !='ZZZ':
    if i==max_i: i=0
    if instructions[i]=='L':
        k = network[k][0] 
    else: k = network[k][1]
    i += 1
    steps += 1

print(steps)

24253


**AI Prompt**

Is it possible to get the values of a dictionary for multiple keys at the same time?

**AI Response**

Yes, there are a few ways to get values for multiple keys from a dictionary:

1. You can use a list comprehension:
```python
[network[k] for k in ['AAA', 'ZZZ']]
```

2. Or use the dictionary's `get()` method for safer access:
```python
[network.get(k) for k in ['AAA', 'ZZZ']]
```

3. Or use `itemgetter` from the operator module:
```python
from operator import itemgetter
itemgetter('AAA', 'ZZZ')(network)
```

Which method would you like to try? Note that all these will return a list/tuple of the values for those keys.

In [ ]:
lines = inp.splitlines()
instructions = lines[0]
network = {}
for line in lines[2:]:
    k,v = line.split('=')
    k = k.strip()
    v = v.strip()
    v1 = v[1:4]
    v2 = v[6:9]
    network[k]=(v1,v2)

keys_list = [k for k in network.keys() if k[2]=='A']
i = 0
steps = 0
max_i=len(instructions)
end = False
while end==False:
    if i==max_i: i=0
    if instructions[i]=='L':
        keys_list =[network[k][0] for k in keys_list] 
    else: keys_list =[network[k][1] for k in keys_list]
    if len([k for k in keys_list if k[2]=='Z'])==len(keys_list): end=True
    i += 1
    steps += 1

print(steps)

---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
Cell In[1], line 22
     20     keys_list =[network[k][0] for k in keys_list] 
     21 else: keys_list =[network[k][1] for k in keys_list]
---> 22 if len([k for k in keys_list if k[2]=='Z'])==len(keys_list): end=True
     23 i += 1
     24 steps += 1

Cell In[1], line 22, in <listcomp>(.0)
     20     keys_list =[network[k][0] for k in keys_list] 
     21 else: keys_list =[network[k][1] for k in keys_list]
---> 22 if len([k for k in keys_list if k[2]=='Z'])==len(keys_list): end=True
     23 i += 1
     24 steps += 1

File /usr/local/lib/python3.11/site-packages/execnb/shell.py:92, in run_cell.<locals>.handler(*args)
---> 92 def handler(*args): raise TimeoutError()

TimeoutError: 


TimeoutError: 

**AI Prompt**

Part 1 was easy but doing part 2 seems to time out. Where is my code not efficient enough to do the job?

**AI Response**

Looking at your code, I can see why it's timing out. You're trying to find when all paths simultaneously end in 'Z', but this could take an extremely long time to compute directly. 

Instead of running until all paths reach 'Z' at the same time, we could:

1. Find the number of steps it takes for each starting position (ending in 'A') to reach its first 'Z' ending position
2. Then use math to find when these cycles align (likely using least common multiple - LCM)

Would you like to try modifying your code to:
1. First track how many steps each individual path takes to reach 'Z'?
2. Then we can work on finding when they align?

## Part Two
The sandstorm is upon you and you aren't any closer to escaping the wasteland. You had the camel follow the instructions, but you've barely left your starting position. It's going to take significantly more steps to escape!

What if the map isn't for people - what if the map is for ghosts? Are ghosts even bound by the laws of spacetime? Only one way to find out.

After examining the maps a bit longer, your attention is drawn to a curious fact: the number of nodes with names ending in A is equal to the number ending in Z! If you were a ghost, you'd probably just start at every node that ends with A and follow all of the paths at the same time until they all simultaneously end up at nodes that end with Z.

For example:
```
LR

11A = (11B, XXX)
11B = (XXX, 11Z)
11Z = (11B, XXX)
22A = (22B, XXX)
22B = (22C, 22C)
22C = (22Z, 22Z)
22Z = (22B, 22B)
XXX = (XXX, XXX)
```
Here, there are two starting nodes, 11A and 22A (because they both end with A). As you follow each left/right instruction, use that instruction to simultaneously navigate away from both nodes you're currently on. Repeat this process until all of the nodes you're currently on end with Z. (If only some of the nodes you're on end with Z, they act like any other node and you continue as normal.) In this example, you would proceed as follows:

Step 0: You are at 11A and 22A.
Step 1: You choose all of the left paths, leading you to 11B and 22B.
Step 2: You choose all of the right paths, leading you to 11Z and 22C.
Step 3: You choose all of the left paths, leading you to 11B and 22Z.
Step 4: You choose all of the right paths, leading you to 11Z and 22B.
Step 5: You choose all of the left paths, leading you to 11B and 22C.
Step 6: You choose all of the right paths, leading you to 11Z and 22Z.
So, in this example, you end up entirely on nodes that end in Z after 6 steps.

Simultaneously start on every node that ends with A. How many steps does it take before you're only on nodes that end with Z?

**AI Prompt**

We have to threat the paths in parallel, the stop condition is whan all paths end at Z at the same time.

**AI Response**

Actually, there's a clever mathematical insight we can use here. Instead of running all paths simultaneously until they align (which could take an extremely long time), we can:

1. Calculate how many steps each individual path takes to reach its first Z
2. Find the least common multiple (LCM) of these numbers

This works because each path forms a cycle - once a path reaches a Z, it will repeat the same pattern.

Would you like to try modifying your code to first find the number of steps for just one path? Then we can extend it to find all paths and use LCM to find when they align.

Would you like to try this approach?

**AI Prompt**

Why should a path cycle after reaching its first key with a Z at the end? The instructions do now show this cycle.

**AI Response**

You raise a good point. Let me explain why we can be confident about the cycling:

1. The network is finite (has a limited number of nodes)
2. The instruction sequence is fixed and repeats
3. At each node, the choices are deterministic (L goes to one specific node, R to another)

This means that once we follow a path:
- Starting from any node
- Following the instruction sequence
- Eventually we MUST repeat a previous (node, instruction_position) combination
- Once we hit such a repeat, we're in a cycle

Would you like to verify this by modifying your code to track and print out the sequence of nodes visited from one starting point? This would help us confirm the cycling behavior.

**AI Prompt**

Yes it will eventually hit a cycle, but a path from the start key that ends with 'A' top the fisrt key that ends with 'Z' is not one cycle.

**AI Response**

Ah, now I understand your concern better. You're right - the path from an 'A' node to a 'Z' node isn't necessarily one cycle. We need to:

1. Find when we first hit a 'Z' node
2. Then continue following that path to understand the actual cycle pattern
3. Make sure we understand if/how the 'Z' nodes appear in the cycle

Would you like to modify your code to track:
- Steps to first 'Z'
- Then continue a bit further to see what happens after that first 'Z'?

This will help us understand the actual pattern better before we make assumptions about cycles.

**AI Prompt**

Lets break down Part 2 of AoC DAy 8 2023 from the beginning. I don't get the thing with the cycle.

**AI Response**

Okay, let's start fresh with Part 2. Here's what we know from the instructions:

1. We need to find all starting nodes that end with 'A'
2. We follow paths from all these nodes simultaneously
3. We're looking for the moment when ALL paths are simultaneously at nodes ending with 'Z'

Let's break it down step by step. Would you like to start by:
1. Finding all the nodes that end with 'A' in your input?
2. Then we can track one path to see what happens when we follow it?

This way we can understand the behavior before trying to solve the whole problem. Would you like to try that?

In [ ]:
keys_list = [k for k in network.keys() if k[2]=='A']
i = 0
steps = 0
max_i=len(instructions)

k = keys_list[3]

while k[2] !='Z':
    if i==max_i: i=0
    if instructions[i]=='L':
        k = network[k][0] 
    else: k = network[k][1]
    i += 1
    steps += 1

print(steps)

16271


In [ ]:
print(keys_list)

['AAA', 'MGA', 'DGA', 'TLA', 'RDA', 'DPA']


**AI Prompt**

Oke done, now what?

**AI Response**

Great! You've found that:
1. You have 6 starting nodes ending in 'A'
2. For one path (starting from 'RDA'), it takes 16271 steps to reach a 'Z' node

Let's understand the pattern better. Could you modify your code to:
1. Find how many steps each of the 6 paths takes to reach their first 'Z' node
2. Store these numbers in a list

This will help us see if there's any pattern in how long each path takes. Would you like to try that?

In [ ]:
keys_list = [k for k in network.keys() if k[2]=='A']
nr_steps=[]
max_i=len(instructions)

for k in keys_list:
    i = 0
    steps = 0
    while k[2] !='Z':
        if i==max_i: i=0
        if instructions[i]=='L':
            k = network[k][0] 
        else: k = network[k][1]
        i += 1
        steps += 1
    nr_steps.append(steps)

print(nr_steps)

[24253, 21797, 14429, 16271, 20569, 13201]


**AI Prompt**

What can I do with this?

**AI Response**

Now we have the number of steps each path takes to reach its first 'Z' node. Since we need all paths to reach 'Z' nodes simultaneously, we need to find the first number that is divisible by all these numbers - this is called the Least Common Multiple (LCM).

Python has an LCM function in the `math` module. Would you like to try:
```python
from math import lcm
answer = lcm(*nr_steps)
```

This should give us the first step number where all paths end at 'Z' nodes simultaneously. Would you like to try that?

In [ ]:
from math import lcm
answer = lcm(*nr_steps)
answer

12357789728873

**AI Prompt**

And this should be the correct answer to part 2?

**AI Response**

Yes! The number 12357789728873 should be the correct answer for Part 2. 

Here's why this works:
1. Each path takes a certain number of steps to reach a 'Z' node (the numbers we found: [24253, 21797, 14429, 16271, 20569, 13201])
2. The LCM gives us the smallest number that's divisible by all these numbers
3. This means it's the first step where all paths will simultaneously end at 'Z' nodes

Would you like to verify this is the correct answer by submitting it? Or would you like to understand more about why the LCM gives us the right answer?

**AI Prompt**

You are right, it is correct, but I still do not get it. IS this network special that each individual path until the first 'Z' is a cycle?

**AI Response**

Actually, I made an assumption earlier that wasn't fully explained. Let's investigate this:

Would you like to modify your code to see what happens AFTER reaching the first 'Z' node for one path? We could:
1. Pick one starting node (say 'AAA')
2. Track the sequence of nodes it visits
3. Continue past the first 'Z' for several more steps
4. Print out this sequence to see if/how it cycles

This would help us understand if the paths are actually cycling, and if so, how. Would you like to try this to verify our assumption about the cycling behavior?

**AI Prompt**

I want to understand why they cycle, it seems obvious that they do, otherwise the solution would not work? Why do they cycle in the given network?

**AI Response**

Ah, let's break down why these paths MUST cycle in this network:

1. The network is finite (has a limited number of nodes)
2. The instruction sequence is fixed and repeats (your L/R sequence)
3. At each node, the choices are deterministic:
   - For 'L', you always go to the same left node
   - For 'R', you always go to the same right node
4. Most importantly: When you're at any node with a specific position in the instruction sequence, your next move is completely determined

This means that if you ever hit the same node at the same position in the instruction sequence, you MUST follow the exact same path afterward. And since:
- There are only a finite number of nodes
- And a finite instruction sequence
- You must eventually hit a combination you've seen before

Would you like to verify this by tracking both the nodes AND the instruction position to see when the cycle starts?

**AI Prompt**

I get it that is must eventually cycle but each path with the individual lengths given in nr_steps need not be cycles, right. But if they are not cycles, then LCM should not work.

**AI Response**

Ah, now I understand your question better! You're right - just knowing that paths eventually cycle isn't enough. For the LCM solution to work, we need an additional property:

For each starting node ending in 'A':
1. The number of steps to first reach 'Z' must be the same as the cycle length
2. After reaching 'Z', it must immediately start the same cycle again

Let's verify this! Would you like to modify the code to check for one path:
1. How many steps to first 'Z'
2. How many steps from that 'Z' to the next 'Z'
3. Whether these numbers are the same

This would help us understand if the input was specially constructed to have this property. Would you like to check this?

**AI Prompt**

Yes it has this property, otherwise the solution would not work, bit that makes the network have a special property which is nowhere stated in the problem?

**AI Response**

Exactly! You've identified something very important - this puzzle relies on a "hidden" property of the input data that isn't mentioned in the problem description. The network was specially constructed so that:

1. Each path from an 'A' node to its first 'Z' node is exactly one cycle length
2. These cycles are "clean" - they start over immediately after reaching Z

This is what makes the LCM solution work, but it's not a property that would be true for any arbitrary network of nodes with the same rules. It's a case where understanding the actual input data's special properties was key to finding the solution.

Would you like to see what would happen in a network that didn't have this special property? We could construct a small example where LCM wouldn't work.

**AI Prompt**

Please summarize the work and give me a nice total solution of Part 2

**AI Response**

Here's a summary of Part 2 and a complete solution:

The puzzle involved finding when multiple paths through a network simultaneously end at nodes ending with 'Z'. The key insights were:
1. Find all starting nodes ending with 'A'
2. The input network was specially constructed so each path's first-Z distance equals its cycle length
3. This allowed us to use LCM (Least Common Multiple) to find when paths align

Here's the complete solution:

```python
from math import lcm

# Parse input
lines = inp.splitlines()
instructions = lines[0]
network = {}
for line in lines[2:]:
    k,v = line.split('=')
    k = k.strip()
    v = v.strip()
    v1 = v[1:4]
    v2 = v[6:9]
    network[k]=(v1,v2)

# Find steps for each starting point
keys_list = [k for k in network.keys() if k[2]=='A']
steps_to_z = []
max_i = len(instructions)

for k in keys_list:
    i = 0
    steps = 0
    while k[2] != 'Z':
        if i == max_i: i = 0
        if instructions[i] == 'L':
            k = network[k][0] 
        else: 
            k = network[k][1]
        i += 1
        steps += 1
    steps_to_z.append(steps)

# Calculate LCM of all path lengths
answer = lcm(*steps_to_z)
print(f"Steps until all paths end at Z: {answer}")
```

Would you like me to explain any part of this in more detail?

**AI Prompt**

Is there a simpler more efficient solution?

**AI Response**

The current solution is already quite efficient in terms of computational complexity. The main operations are:

1. Parsing the input (one pass)
2. Finding the steps for each starting point (we only need to go until first Z)
3. Computing LCM (built-in Python function)

The only potential optimization might be combining steps 1 and 2 into a single pass through the input, but that would make the code less readable and wouldn't significantly improve performance.

The solution relies on the special property of the input (that paths form perfect cycles from first Z), which allows us to use LCM. Without this property, we'd need a much more complex solution.

Would you like to see any specific part optimized, or are you wondering about a different approach entirely?

**AI Prompt**

Give the best solution for Part 2 that you can come up with.

**AI Response**

Here's a concise and efficient solution for Part 2, with minimal overhead:

```python
from math import lcm

# Parse and solve in one pass
instructions = inp.splitlines()[0]
network = {}
starts = []

# Parse network and collect starting points
for line in inp.splitlines()[2:]:
    node = line[0:3]
    if node[2] == 'A':
        starts.append(node)
    network[node] = (line[7:10], line[12:15])

# Find cycles for all starting points
cycles = []
for pos in starts:
    steps = 0
    i = 0
    while pos[2] != 'Z':
        pos = network[pos][0 if instructions[i] == 'L' else 1]
        steps += 1
        i = (i + 1) % len(instructions)
    cycles.append(steps)

print(lcm(*cycles))
```

Key improvements:
1. Single pass through input for parsing and finding start nodes
2. More efficient string slicing instead of split/strip
3. Modulo operator for instruction index instead of if-reset
4. Minimal variable creation and memory usage

Would you like me to explain any of these optimizations in detail?